src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersCodeFixProvider.cs

```diff
@@ -0,0 +1,126 @@
+    internal abstract class AbstractRemoveUnusedMembersCodeFixProvider<TFieldDeclarationSyntax> : SyntaxEditorBasedCodeFixProvider
+        where TFieldDeclarationSyntax : SyntaxNode
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(IDEDiagnosticIds.RemoveUnusedMembersDiagnosticId);
+
+        // Adjust declarators to remove based on whether or not all variable declarators within a field declaration should be removed.
+        protected abstract void AdjustDeclarators(HashSet<TFieldDeclarationSyntax> fieldDeclarators, HashSet<SyntaxNode> declarators);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(new MyCodeAction(
+                c => FixAsync(context.Document, context.Diagnostics[0], c)),
+                context.Diagnostics);
+            return Task.CompletedTask;
+        }
+
+        protected override async Task FixAllAsync(
+            Document document,
+            ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor,
+            CancellationToken cancellationToken)
+        {
+            var declarators = new HashSet<SyntaxNode>();
+            var fieldDeclarators = new HashSet<TFieldDeclarationSyntax>();
+
+            // Compute declarators to remove, and also track common field declarators.
+            foreach (var diagnostic in diagnostics)
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                // Get symbol to be removed.
+                var diagnosticNode = diagnostic.Location.FindNode(getInnermostNodeForTie: true, cancellationToken);
+                var symbol = semanticModel.GetDeclaredSymbol(diagnosticNode, cancellationToken);
+                Debug.Assert(symbol != null);
+
+                // Get symbol declarations to be removed.
+                var declarationService = document.GetLanguageService<ISymbolDeclarationService>();
+                foreach (var declReference in declarationService.GetDeclarations(symbol))
+                {
+                    var node = await declReference.GetSyntaxAsync(cancellationToken).ConfigureAwait(false);
+                    declarators.Add(node);
+
+                    // For fields, the declaration node is the variable declarator.
+                    // We also track the ancestor FieldDeclarationSyntax which may declare more then one field.
+                    if (symbol.Kind == SymbolKind.Field)
+                    {
+                        var fieldDeclarator = node.FirstAncestorOrSelf<TFieldDeclarationSyntax>();
+                        fieldDeclarators.Add(fieldDeclarator);
+                    }
+                }
+            }
+
+            // If all the fields declared within a field declaration are unused,
+            // we can remove the entire field declaration instead of individual variable declarators.
```

Comment:
ah, that makes sense.  woudl perhaps be good to explain this above on the helper method.

Reply:
Done.

---
In reply to: [213410938](https://github.com/dotnet/roslyn/pull/29511#discussion_r213410938) [](ancestors = 213410938)